File: C++/interviewAlgorithms/findNearestNumber.cpp
#include<iostream>
#include<vector>
using namespace std;

class FindNearestNumber{
    public:
        static vector<int> findNearestNumber(vector<int>& numbers){
            //1.从后向前查看逆序区域，找到逆序区域的前一位，也就是数字置换的边界
            int index = findTransferPoint(numbers);
            //如果数字置换边界是0，说明整个数组已经逆序，无法得到更大的相同数字组成的整数，返回空向量
            if(index == 0){
                return {};
            }
            //2.把逆序区域前一位和逆序区域中刚刚大于它的数字交换位置
            //复制并入参，避免直接修改入参
            vector<int> numbersCopy = numbers;
            exchangeHead(numbersCopy,index);
            //3.把原来的逆序区域转为顺序
            reverse(numbersCopy,index);
            return numbersCopy;
        }
        static void outputNumbers(vector<int>& numbers){
            for(int i: numbers){
                cout<<i<<"";
            }
            cout<<endl;
        }

    private:
        static int findTransferPoint(vector<int>& numbers){
            int length = numbers.size();
            for(int i =length-1;i>0;i--){
                if(numbers[i] > numbers[i-1]){
                    return i;
                }
            }
            return 0;
        }
        static void exchangeHead(vector<int>& numbers,int index){
            int head = numbers[index-1];
            for(int i =numbers.size()-1;i>0;i--){
                if(head<numbers[i]){
                    numbers[index-1] = numbers[i];
                    numbers[i] = head;
                    break;
                }
            }
        }
        static void reverse(vector<int>& numbers,int index){
            int i =index,j=numbers.size()-1;
            while(i<j){
                swap(numbers[i], numbers[j]);
                i++;
                j--;
            }
        }
};

int main(){
    vector<int> numbers = {1,2,3,4,5};
    int length = numbers.size();
    for(int i =0; i<10;i++){
        numbers = FindNearestNumber::findNearestNumber(numbers);
        FindNearestNumber::outputNumbers(numbers);
    }
    return 0;
}

File: python/interviewAlgorithms/findNearestNumber.py
class FindNearestNumber:
    @staticmethod
    def findNearestNumber(numbers):
        # 1. 从后向前查看逆序区域，找到逆序区域的前一位，也就是数字置换的边界
        index = FindNearestNumber.findTransferPoint(numbers)
        # 如果数字置换边界是0，说明整个数组已经逆序，无法得到更大的相同数字组成的整数，返回None
        if index == 0:
            return None
        
        # 2. 如果把逆序区域的前一位和逆序区域中刚刚大于它的数字交换位置，复制并入参，避免直接修改入参
        numbers_copy = numbers.copy()
        numbers_copy = FindNearestNumber.exchangeHead(numbers_copy, index)
        # 3. 将逆序区域翻转，使其变为顺序
        numbers_copy = FindNearestNumber.reverse(numbers_copy, index)
        return numbers_copy

    @staticmethod
    def findTransferPoint(numbers):
        for i in range(len(numbers)-1, 0, -1):
            if numbers[i] > numbers[i-1]:
                return i
        
        return 0 
    
    @staticmethod
    def exchangeHead(numbers, index):
        head = numbers[index-1]
        for i in range(len(numbers)-1, 0, -1):
            if head < numbers[i]:
                numbers[index-1], numbers[i] = numbers[i], head
                break
        return numbers
    
    @staticmethod
    def reverse(numbers, index):
        i, j = index, len(numbers)-1
        while i < j:
            temp = numbers[i]
            numbers[i] = numbers[j]
            numbers[j] = temp
            i += 1
            j -= 1
        return numbers

    @staticmethod
    def outputNumbers(numbers):
        for i in numbers:
            print(i)

if __name__ == "__main__":
    numbers = [1,2,3,4,5]
    # 打印12345之后的10个全排列整数
    for i in range(10):
        numbers = FindNearestNumber.findNearestNumber(numbers)
        if numbers is None:
            break
        FindNearestNumber.outputNumbers(numbers)
        print()  
